3277: Migrate from styled-components to emotion

steffenkleinle opened the pull request:
### Short Description

<!-- Describe this PR in one or two sentences. -->
Migrate from styled-components to emotion to prepare for design system implementation (and due to styled-components being maintenance mode).

### Proposed Changes

<!-- Describe this PR in more detail. -->

- Migrate from styled-components to emotion
- Remove `transient` props as its no emotion feature (props prefixed with `$`)

### Side Effects

<!-- List all related components that have not been explicitly changed but may be affected by this PR -->

Hopefully none.

### Testing

<!-- List all things that should be tested while reviewing this PR. -->
Please test everything thoroughly as I had to do quite a few changes.

### Resolved Issues

<!-- List all issues which should be closed when this PR is merged. -->

Fixes: N/A

---

<!--
DOR:
- [Release notes](https://github.com/digitalfabrik/integreat-app/blob/main/docs/contributing.md#release-notes) have been added for user visible changes
- Linting: `yarn lint`
- Typescript: `yarn ts:check`
- Prettier: `yarn prettier`
- Unit tests: `yarn test`
-->

bahaaTuffaha commented:
> Due to https://github.com/digitalfabrik/react-sticky-headroom requiring styled-components as a peer dependency, its not yet possible to remove styled-components for web. Otherwise hopefully none.

What if we used patch-package on react-sticky-headroom  just to replace styled-components 🤔?

steffenkleinle commented:
> > Due to https://github.com/digitalfabrik/react-sticky-headroom requiring styled-components as a peer dependency, its not yet possible to remove styled-components for web. Otherwise hopefully none.
> 
> What if we used patch-package on react-sticky-headroom just to replace styled-components 🤔?

We maintain the package ourselves, so I will fix it and create a new release. Will do that in a separate PR though.

bahaaTuffaha commented:
There is transient props left here:

https://github.com/digitalfabrik/integreat-app/blob/17aeed794694ae982884ed7a3f5948efa0ea9c2c/web/src/components/SharingPopup.tsx#L23-L26

bahaaTuffaha reviewed (commented):
Found a bug not sure if its related.. if you keep pressing the button the icon disappear: 

https://github.com/user-attachments/assets/9e76d354-e0e2-4d04-bc4c-23891ac52115

bahaaTuffaha reviewed (changes requested):
go to http://localhost:9000/budel-cranendonck/de

```
Uncaught Error: Component selectors can only be used in conjunction with @emotion/babel-plugin, the swc Emotion plugin, or another Emotion-aware compiler transform.
    at handleInterpolation (emotion-serialize.development.esm.js:88:1)
    at serializeStyles (emotion-serialize.development.esm.js:267:1)
    at css (emotion-react.browser.development.esm.js:446:25)
    at Selector.tsx:68:8
  
  ```
    
    I think you need to adjust `babel.config.js` or something.

steffenkleinle commented on web/src/components/base/Button.tsx:
```diff
@@ -1,10 +1,10 @@
-  cursor: ${props => (props.$disabled ? 'default' : 'pointer')};
-  pointer-events: ${props => (props.$disabled ? 'none' : 'default')};
+  cursor: ${props => (props.disabled ? 'default' : 'pointer')};
+  pointer-events: ${props => (props.disabled ? 'none' : 'default')};
```
Since this is not part of this PR and it will probably be obsolete with mui anyway, I'll keep this as it is. In general good suggestion though.

bahaaTuffaha commented on web/src/components/PoisMobile.tsx:
```diff
@@ -27,11 +27,11 @@ const ListTitle = styled.div`
 `
 
-const GoBackContainer = styled.div<{ $hidden: boolean }>`
+const GoBackContainer = styled.div<{ hidden: boolean }>`
```
we can customize prop forwarding and create your own styled (not tested):
```js
import { styled as emotionStyled } from '@emotion/styled';
import isPropValid from '@emotion/is-prop-valid';

const styled = (component) =>
  emotionStyled(component, {
    shouldForwardProp: (prop) => isPropValid(prop) && !prop.startsWith('$'),
  });

export default styled;
```

steffenkleinle commented on web/src/components/ChatMessage.tsx:
```diff
@@ -22,15 +22,15 @@ export const Message = styled.div`
+const Container = styled.div<{ isAuthor: boolean }>`
   display: flex;
-  ${props => (props.$isAuthor ? 'flex-direction:row-reverse' : 'flex-direction:row')};
+  ${props => (props.isAuthor ? 'flex-direction:row-reverse' : 'flex-direction:row')};
```
None besides the fact that it was not part of this PR :)

steffenkleinle commented on web/src/components/Footer.tsx:
```diff
@@ -8,20 +8,19 @@ type FooterProps = {
-    margin: ${props => (props.$overlay ? 0 : '5px')};
-    color: ${props => props.theme.isContrastTheme && !props.$overlay && props.theme.colors.textColor};
+    margin: ${props => (props.overlay ? 0 : '5px')};
+    color: ${props => props.theme.isContrastTheme && !props.overlay && props.theme.colors.textColor};
```
Not part of this PR, might be refactored later.

steffenkleinle commented on web/src/components/base/Icon.tsx:
```diff
@@ -1,10 +1,11 @@
-import styled from 'styled-components'
 
-const StyledIcon = styled(SVG)<{ $directionDependent: boolean; $reverse: boolean }>`
+const StyledIcon = styled(SVG, { shouldForwardProp })<{ directionDependent: boolean; reverse: boolean }>`
```
As discussed: Since we only have two occurrences of this, we'll keep this as it is for now.

LeandraH reviewed (approved):
Looks great!

steffenkleinle commented:
> ❓ I think the integreat-chat needed some adjustment not sure why did you remove the height checking for the input field.

Done, I added `height: 100%;` as per Leandra's suggestion. I removed the height logic as it was overly complicated and error prone and could easily be replaced with `space-between`.
